src/common/fanctl.cpp

```diff
@@ -0,0 +1,182 @@
+// fanctl.cpp
+
+#include "fanctl.h"
+#include "stm32f4xx_hal.h"
+#include "gpio.h"
+
+// global variables for C wrapper
+static uint8_t CFanCtl_count = 0;                  // number of instances
+static CFanCtl *CFanCtl_instance[FANCTL_MAX_FANS]; // array of pointers to instances
+
+//------------------------------------------------------------------------------
+// CFanCtlPWM implementation
+
+CFanCtlPWM::CFanCtlPWM(uint8_t pin_out, uint8_t pwm_min, uint8_t pwm_max) {
+    pin = pin_out;
+    min_value = pwm_min;
+    max_value = pwm_max;
+    initialized = false;
+    output_state = false;
+    pha_ena = false;
+    pwm = 0;
+    cnt = 0;
+    val = 0;
+    pha = 0;
+    pha_max = 0;
+    pha_stp = 0;
+}
+
+void CFanCtlPWM::init() {
+    gpio_set(pin, 0);
+    gpio_init(pin, GPIO_MODE_OUTPUT_PP, GPIO_NOPULL, GPIO_SPEED_FREQ_HIGH);
+    initialized = true;
+}
+
+int8_t CFanCtlPWM::tick() {
+    if (!initialized)          // ?not initialised
+        return -1;             // return -1 means pwm output is off
+    int8_t pwm_on = cnt - pha; // calculate on time (number of ticks after 0-1 pwm transition)
+    if (pwm_on >= val)
+        pwm_on -= max_value;
+    bool o = (cnt >= pha) && (cnt < (pha + val));
+    if (++cnt >= max_value) {
+        cnt = 0;
+        if (val != pwm) { // pwm changed
+            val = pwm;    // update cached value
+            pha = 0;      // reset phase
+            if ((val > 1) && (val <= (max_value / 2))) {
+                pha_max = max_value - val;
+                uint8_t steps = max_value / val; // calculate number of steps
+                if (steps < 3)
+                    steps = 3;               // limit steps >= 3
```

Comment:
```suggestion
                const uint8_t steps = std::max(3, max_value / val); // calculate number of steps (at least 3)
```

Reply:
@espr14 hmmm, this may produce unnecessarily larger and slower code on the 8-bit platform: std::max will take ints (aka 16bit) to compute the result, which is 2x more data than necessary.

Reply:
I agree with @DRracer